weaviate/batch/crud_batch.py

```diff
@@ -672,41 +750,164 @@ def create_references(self) -> list:
             return response.json()
         return []
 
+    def _flush_in_thread(self,
+            data_type: str,
+            batch_request: BatchRequest,
+            ) -> Tuple[Response, int]:
+        """
+        Flush BatchRequest in current thread/process.
+
+        Parameters
+        ----------
+        data_type : str
+            The data type of the BatchRequest, used to save time for not checking the type of the
+            BatchRequest.
+        batch_request : weaviate.batch.BatchRequest
+            Contains all the data objects that should be added in one batch.
+            Note: Should be a sub-class of BatchRequest since BatchRequest
+            is just an abstract class, e.g. ObjectsBatchRequest, ReferenceBatchRequest
+
+        Returns
+        -------
+        Tuple[requests.Response, int]
+            The request response and number of items sent with the BatchRequest as tuple.
+        """
+
+        if len(batch_request) != 0:
+            response = self._create_data(
+                data_type=data_type,
+                batch_request=batch_request,
+            )
+            return response, len(batch_request)
+        return None, 0
+
+    def _send_batch_requests(self, force_wait: bool) -> None:
+        """
+        Send BatchRequest in a separate thread/process. This methods submits a task to create only
+        the ObjectsBatchRequests to the BatchExecutor and adds the ReferencesBatchRequests to a
+        queue, then it carries on in the main thread until `num_workers` tasks have been submitted.
+        When we have reached number of tasks to be equal to `num_workers` it waits for all the
+        tasks to finish and handles the responses. After all ObjectsBatchRequests have been handled
+        it created separate tasks for each ReferencesBatchRequests, then it handles their responses
+        as well. This mechanism of creating References after Objects is constructed in this manner
+        to eliminate potential error when creating references from a object that does not yet
+        exists (object that is part of another task).
+
+        Parameters
+        ----------
+        force_wait : bool
+            Whether to wait on all created tasks even if we do not have `num_workers` tasks created
+        """
+
+        future = self._executor.submit(
+            self._flush_in_thread,
+            data_type='objects',
+            batch_request=self._objects_batch,
+        )
+        self._future_pool.append(future)
+        if len(self._reference_batch) != 0:
+            self._reference_batch_queue.append(
+                self._reference_batch
+            )
+        self._objects_batch = ObjectsBatchRequest()
+        self._reference_batch = ReferenceBatchRequest()
+
+        if not force_wait and len(self._future_pool) < self._num_workers:
+            return
+
+        creation_time = self._connection.timeout_config[1] / self._num_workers
```

Comment:
as none of this changes during runtime, maybe move this to the __init__/ configure function and have it as a member?

Reply:
Will do